// effect/dist/esm/internal/stm/tPubSub.js
import * as RA from "../../Array.js";
import * as Effect from "../../Effect.js";
import { dual, identity, pipe } from "../../Function.js";
import * as HashSet from "../../HashSet.js";
import * as Option from "../../Option.js";
import * as core from "./core.js";
import * as OpCodes from "./opCodes/strategy.js";
import * as stm from "./stm.js";
import * as tQueue from "./tQueue.js";
import * as tRef from "./tRef.js";
/** @internal */
const TPubSubSymbolKey = "effect/TPubSub";
/** @internal */
export const TPubSubTypeId = /*#__PURE__*/Symbol.for(TPubSubSymbolKey);
const AbsentValue = /*#__PURE__*/Symbol.for("effect/TPubSub/AbsentValue");
/** @internal */
export const makeNode = (head, subscribers, tail) => ({
  head,
  subscribers,
  tail
});
/** @internal */
class TPubSubImpl {
  pubsubSize;
  publisherHead;
  publisherTail;
  requestedCapacity;
  strategy;
  subscriberCount;
  subscribers;
  [TPubSubTypeId] = {
    _A: _ => _
  };
  [tQueue.TEnqueueTypeId] = tQueue.tEnqueueVariance;
  constructor(pubsubSize, publisherHead, publisherTail, requestedCapacity, strategy, subscriberCount, subscribers) {
    this.pubsubSize = pubsubSize;
    this.publisherHead = publisherHead;
    this.publisherTail = publisherTail;
    this.requestedCapacity = requestedCapacity;
    this.strategy = strategy;
    this.subscriberCount = subscriberCount;
    this.subscribers = subscribers;
  }
  isShutdown = /*#__PURE__*/core.effect(journal => {
    const currentPublisherTail = tRef.unsafeGet(this.publisherTail, journal);
    return currentPublisherTail === undefined;
  });
  awaitShutdown = /*#__PURE__*/core.flatMap(this.isShutdown, isShutdown => isShutdown ? stm.void : core.retry);
  capacity() {
    return this.requestedCapacity;
  }
  size = /*#__PURE__*/core.withSTMRuntime(runtime => {
    const currentPublisherTail = tRef.unsafeGet(this.publisherTail, runtime.journal);
    if (currentPublisherTail === undefined) {
      return core.interruptAs(runtime.fiberId);
    }
    return core.succeed(tRef.unsafeGet(this.pubsubSize, runtime.journal));
  });
  isEmpty = /*#__PURE__*/core.map(this.size, size => size === 0);
  isFull = /*#__PURE__*/core.map(this.size, size => size === this.capacity());
  offer(value) {
    return core.withSTMRuntime(runtime => {
      const currentPublisherTail = tRef.unsafeGet(this.publisherTail, runtime.journal);
      if (currentPublisherTail === undefined) {
        return core.interruptAs(runtime.fiberId);
      }
      const currentSubscriberCount = tRef.unsafeGet(this.subscriberCount, runtime.journal);
      if (currentSubscriberCount === 0) {
        return core.succeed(true);
      }
      const currentPubSubSize = tRef.unsafeGet(this.pubsubSize, runtime.journal);
      if (currentPubSubSize < this.requestedCapacity) {
        const updatedPublisherTail = new tRef.TRefImpl(void 0);
        const updatedNode = makeNode(value, currentSubscriberCount, updatedPublisherTail);
        tRef.unsafeSet(currentPublisherTail, updatedNode, runtime.journal);
        tRef.unsafeSet(this.publisherTail, updatedPublisherTail, runtime.journal);
        tRef.unsafeSet(this.pubsubSize, currentPubSubSize + 1, runtime.journal);
        return core.succeed(true);
      }
      switch (this.strategy._tag) {
        case OpCodes.OP_BACKPRESSURE_STRATEGY:
          {
            return core.retry;
          }
        case OpCodes.OP_DROPPING_STRATEGY:
          {
            return core.succeed(false);
          }
        case OpCodes.OP_SLIDING_STRATEGY:
          {
            if (this.requestedCapacity > 0) {
              let currentPublisherHead = tRef.unsafeGet(this.publisherHead, runtime.journal);
              let loop = true;
              while (loop) {
                const node = tRef.unsafeGet(currentPublisherHead, runtime.journal);
                if (node === undefined) {
                  return core.retry;
                }
                const head = node.head;
                const tail = node.tail;
                if (head !== AbsentValue) {
                  const updatedNode = makeNode(AbsentValue, node.subscribers, node.tail);
                  tRef.unsafeSet(currentPublisherHead, updatedNode, runtime.journal);
                  tRef.unsafeSet(this.publisherHead, tail, runtime.journal);
                  loop = false;
                } else {
                  currentPublisherHead = tail;
                }
              }
            }
            const updatedPublisherTail = new tRef.TRefImpl(void 0);
            const updatedNode = makeNode(value, currentSubscriberCount, updatedPublisherTail);
            tRef.unsafeSet(currentPublisherTail, updatedNode, runtime.journal);
            tRef.unsafeSet(this.publisherTail, updatedPublisherTail, runtime.journal);
            return core.succeed(true);
          }
      }
    });
  }
  offerAll(iterable) {
    return core.map(stm.forEach(iterable, a => this.offer(a)), RA.every(identity));
  }
  shutdown = /*#__PURE__*/core.effect(journal => {
    const currentPublisherTail = tRef.unsafeGet(this.publisherTail, journal);
    if (currentPublisherTail !== undefined) {
      tRef.unsafeSet(this.publisherTail, void 0, journal);
      const currentSubscribers = tRef.unsafeGet(this.subscribers, journal);
      HashSet.forEach(currentSubscribers, subscriber => {
        tRef.unsafeSet(subscriber, void 0, journal);
      });
      tRef.unsafeSet(this.subscribers, HashSet.empty(), journal);
    }
  });
}
/** @internal */
class TPubSubSubscriptionImpl {
  pubsubSize;
  publisherHead;
  requestedCapacity;
  subscriberHead;
  subscriberCount;
  subscribers;
  [TPubSubTypeId] = TPubSubTypeId;
  [tQueue.TDequeueTypeId] = tQueue.tDequeueVariance;
  constructor(pubsubSize, publisherHead, requestedCapacity, subscriberHead, subscriberCount, subscribers) {
    this.pubsubSize = pubsubSize;
    this.publisherHead = publisherHead;
    this.requestedCapacity = requestedCapacity;
    this.subscriberHead = subscriberHead;
    this.subscriberCount = subscriberCount;
    this.subscribers = subscribers;
  }
  isShutdown = /*#__PURE__*/core.effect(journal => {
    const currentSubscriberHead = tRef.unsafeGet(this.subscriberHead, journal);
    return currentSubscriberHead === undefined;
  });
  awaitShutdown = /*#__PURE__*/core.flatMap(this.isShutdown, isShutdown => isShutdown ? stm.void : core.retry);
  capacity() {
    return this.requestedCapacity;
  }
  size = /*#__PURE__*/core.withSTMRuntime(runtime => {
    let currentSubscriberHead = tRef.unsafeGet(this.subscriberHead, runtime.journal);
    if (currentSubscriberHead === undefined) {
      return core.interruptAs(runtime.fiberId);
    }
    let loop = true;
    let size = 0;
    while (loop) {
      const node = tRef.unsafeGet(currentSubscriberHead, runtime.journal);
      if (node === undefined) {
        loop = false;
      } else {
        const head = node.head;
        const tail = node.tail;
        if (head !== AbsentValue) {
          size = size + 1;
          if (size >= Number.MAX_SAFE_INTEGER) {
            loop = false;
          }
        }
        currentSubscriberHead = tail;
      }
    }
    return core.succeed(size);
  });
  isEmpty = /*#__PURE__*/core.map(this.size, size => size === 0);
  isFull = /*#__PURE__*/core.map(this.size, size => size === this.capacity());
  peek = /*#__PURE__*/core.withSTMRuntime(runtime => {
    let currentSubscriberHead = tRef.unsafeGet(this.subscriberHead, runtime.journal);
    if (currentSubscriberHead === undefined) {
      return core.interruptAs(runtime.fiberId);
    }
    let value = AbsentValue;
    let loop = true;
    while (loop) {
      const node = tRef.unsafeGet(currentSubscriberHead, runtime.journal);
      if (node === undefined) {
        return core.retry;
      }
      const head = node.head;
      const tail = node.tail;
      if (head !== AbsentValue) {
        value = head;
        loop = false;
      } else {
        currentSubscriberHead = tail;
      }
    }
    return core.succeed(value);
  });
  peekOption = /*#__PURE__*/core.withSTMRuntime(runtime => {
    let currentSubscriberHead = tRef.unsafeGet(this.subscriberHead, runtime.journal);
    if (currentSubscriberHead === undefined) {
   ùUé  ‹EÔÇ@'   ÆEüMğè³âÿÆEü‹Eì‰Eä…Àt‹uè€¾X   uèâ¿Q‹Eä‰†$   ‹EÔ‹Èjj‰Eğè9ûïÿ‹Mà‹ÿô  ‹ğEÜPÿÔQTÿuàÿuÜÿ¶$  ÿ¶   ‹uğ‹ÎèÖ~ ‹EV‹HX‹ÿ4  ‹ğ…öˆFÿÿÿ‹E‹ˆè  …Ét‹I8…Ét‹ÿuğÿ0  ‹ğ…öˆÿÿÿ‹O,‹ÿP`‹uğ‹ÎPjè¯úïÿ‹‹ÏÿPP‹ …Àt3ŠPûöÂ`u‹È‰Eë‹‰M…ÉtöÂ@u‹@üÑèëÿ0ÿp®[‹MÑèë3É3ÀPQj‹Îè°^÷ÿ‹ğ…öˆ¯şÿÿ‹‹ÏÿPH‹ …Àt(ŠHû‹ğöÁ`t‹0…ötöÁ@u‹@üÑèëÿ0ÿp®[Ñèë3ö3À‹MğPVj
èa^÷ÿ‹ğ…öˆ`şÿÿ‹uğ‹Îjè+›÷ÿ„Àtv‹O,UàR‹ÿì   ÆEü‹ …Àt3ŠPûöÂ`u‹È‰Eë‹‰M…ÉtöÂ@u‹@üÑèëÿ0ÿp®[‹MÑèë3É3ÀPQj‹Îèì]÷ÿ‹ğÆEüMàj è|GŞÿ…öˆİıÿÿ‹uğj‹Îè¨š÷ÿ„Àt‹O,V‹ÿ  ‹ğ…öˆ·ıÿÿ‹uğj‹Îè‚š÷ÿ„Àtv‹‹ÏÿØ   …Àth‹‹ÏÿØ   ‹H,…ÉtW‹ÿPH‹ …Àt3ŠPûöÂ`u‹È‰Eë‹‰M…ÉtöÂ@u‹@üÑèëÿ0ÿp®[‹MÑèë3É3ÀPQj‹Îè2]÷ÿ‹ğ…öˆ1ıÿÿMÔè £üÿÆEü ‹EÔ…Àt‹MØj PèjöU‹‹ÏÿÜ   …Àt@‹‹ÏÿÜ   ‹p8…öt/‹‹ÎÿL  …Àt!‹‹ÎÿL  ‹O,Á¨   Q‹Èèb‘ ‹ğ…öx3öÇEüÿÿÿÿ‹KÿuĞ‹ÿPD‹Æ‹Môd‰
    Y_^[‹å]Â h•\ha  hÀ\èšİÿÌÌÌÌÌÌÌÌÌÌÌÌÌÌfU‹ìjÿhCd¡    PƒìSVW¡ØŒ3ÅPEôd£    ‰eğ‹ù‰}ä3ö‰uì‰uü‹]Sèíúÿÿ‰EèÇEüÿÿÿÿë!èìßÿ‰Eè¸Ío ÃÇEüÿÿÿÿ‹}ä‹]‹uì‹Eè…ÀEğ‰uìÇEü   jS‹ÏèÇc  ‰EàÇEüÿÿÿÿëè¦ßÿ‰Eà¸ o ÃÇEüÿÿÿÿ‹}ä‹uì‹Eà…Àt…öx
…Àx…öu‹ğ‰EìÇEü   ‹O,è‰rûÿ‰EÜÇEüÿÿÿÿëèXßÿ‰EÜ¸a o ÃÇEüÿÿÿÿ‹uì‹EÜ…Àt…öx…Àx…öt‹Æ‹Môd‰
    Y_^[‹å]Â ÌÌÌÌÌÌÌÌÌÌÌÌfU‹ìjÿhCd¡    Pƒì8SVW¡ØŒ3ÅPEôd£    ‹ñ‰uì¡”WÇEÀ    ÿpÿøQT‰E¼…Àtxë3ÿ‹O‰}À‹ÿP@‹Ø‰]ÄÇEü    ‹Î‹‹F,‰Eèÿ’  ‹Eö€u  „   €=èÆZ   ‹@XÇEÌ    ‰EØÇEÈ    ÇEÌHÆZÆEü‹
|«^d¡,   ‹4ˆ‰uğ€¾X   uèªºQ€¾X   ‹†$   ‰Eàuè“ºQÇ†$   ÀcÆEüMäèyúÿÆEüEÈ‹MÌPè®åU‹ğ…öyVÆEüMäè\­âÿÆEü‹Eà‰E…Àt‹Eğ€¸X   uè<ºQ‹Eğ‹M‰ˆ$   ÆEü ‹EÈ…À„  ‹MÌj Pè7óUé}  ‹EÈÇ@'   ÆEüMäèü¬âÿÆEü‹Eà‰EÔ…Àt‹uğ€¾X   uèÜ¹Q‹EÔ‰†$   ‹MÈ3À9E‰Mğ•ÀE   Pjè'õïÿ‹MØ‹ÿô  ‹ğEĞPÿÔQTÿuÔ‹MğÿuĞÿ¶$  ÿ¶   èÆx ‹MìUÜRÿuWÀÇEÜ  ‹fEĞÿ8  ‹ğ…öˆ*ÿÿÿ‹EÜ-  t ƒètƒèu¸   ë‹MÔ‹EĞë¸   ë¸   3É‹uğQPj
‹Îè\–ûÿ‹Mè‹ÿP`Pj‹Îèzôïÿ‹Mì‹ÿPP‹ …Àt3ŠPûöÂ`u‹È‰Eë‹‰M…ÉtöÂ@u‹@üÑèëÿ0ÿp®[‹MÑèë3É3ÀPQj‹ÎèzX÷ÿ‹ğ…öˆşÿÿ‹Mì‹ÿPH‹ …Àt(ŠHû‹ğöÁ`t‹0…ötöÁ@u‹@üÑèëÿ0ÿp®[Ñèë3ö3À‹MğPVj
è*X÷ÿ‹ğ…öˆ/şÿÿ‹uğ‹Îjèô”÷ÿ„Àtv‹MèUÔR‹ÿì   ÆEü‹ …Àt3ŠPûöÂ`u‹È‰Eë‹‰M…ÉtöÂ@u‹@üÑèëÿ0ÿp®[‹MÑèë3É3ÀPQj‹ÎèµW÷ÿ‹ğÆEüMÔj èEAŞÿ…öˆ¬ıÿÿ‹uğj‹Îèq”÷ÿ„Àt‹MèV‹ÿ  ‹ğ…öˆ†ıÿÿ‹uğ‹MØV‹ÿ4  ‹ğ…öˆmıÿÿ‹E‹ˆè  …Ét‹I8…Ét‹ÿuğÿ0  ‹ğ…öˆDıÿÿ‹Mğjè”÷ÿ‹uì„Àto‹‹ÎÿØ   …Àta‹‹ÎÿØ   ‹H,…ÉtP‹ÿPH‹ …Àt(ŠHû‹ğöÁ`t‹0…ötöÁ@u‹@üÑèëÿ0ÿp®[Ñèë3ö3À‹MğPVjèÂV÷ÿ‹ğ…öˆÇüÿÿ‹uìMÈè­œüÿÆEü ‹EÈ…Àt‹MÌj Pè÷ïU‹‹ÎÿÜ   ‹p8…öt/‹‹ÎÿL  …Àt!‹‹ÎÿL  ‹MèÁ¨   Q‹ÈèıŠ ‹ğ…öx3öÇEüÿÿÿÿ‹OS‹ÿPD‹Æ‹Môd‰
    Y_^[‹å]Â ÌÌÌÌÌÌÌÌÌÌÌÌÌÌÌƒìD$V‰D$‹ñD$‰t$Pè¤pøÿ‹N,ÿt$‹ÿü   ^ƒÄÂ ÌÌÌÌÌÌÌÌÌÌÌÌÌÌ‹I,‹ÿ ü   ÌÌÌÌÌfU‹ìjÿhÍCd¡    PƒìVW¡ØŒ3ÅPEôd£    ‹ùÿuè#øÿ‹÷…ÿt%€    ‹‹ÎÿP`=¢† tf‹‹ÎÿÌ   ‹ğ…öuâ‹Y‹Ê‹‚Ğ   …ÀEÈ‹h  ƒàƒÈ ti‹÷…ÿ„•   €    ‹‹ÎÿP`= † „|   ‹‹ÎÿÌ   ‹ğ…öuŞëj‹÷‹‹ÎÿP`=¢† t‹‹ÎÿÌ   ‹ğ…öuâ‹†P  -t  t[ƒèvtVëB‹Š,  ‹ÿP‹ğ…öu%‹÷…ÿt‹‹ÎÿP`= † t‹‹ÎÿÌ   ‹ğ…öuâ¾°  Æ  uhæ  èlZèÿ…À„Œ   ¿P  ê  te‹E‹@X‰EìÇEğ    ÇEü   Eì‰EäEğ‰EèEäPèü‚øÿEğ‰}è‰EäEäPèzoøÿÇEüÿÿÿÿ‹Eğ…Àt4‹PÿQ‹Môd‰
    Y_^‹å]Â ƒ¿l   tE‰}ä‰EèEäPèè€øÿ‹Môd‰
    Y_^‹å]Â ÌÌÌÌÌf‹ÿd  3É…ÀIÁÃÌÌÌÌÌÌÌÌÌÌÌÌÌÌfU‹ìjÿhuFd¡    PƒìHSVW¡ØŒ3ÅPEôd£    ‹Ù‹ÿd  ‹ø…ÿˆé
  ‹K,‰MÜ¹,  Ğ  uZ‹Y‹Â‹²Ğ   …öEÆ‹€h  %   ƒÈ u78‚Ç¬ u/…ö‹ÂEÆ‹€h  %   ƒÈ t	€ºÈ¬  uè°  ‹ø…ÿˆ}
  ¡”WÇEÔ    ÿpÿøQT‰EĞ…Àtxë3ÿ‹O‰}ì‰}Ô‹ÿP@‰EØÇEü    ‹OQ‹‹r$‹Ôh  ÇÿÿÿÿÿÖ‹ğ…ö„4
  j j M¬èMëßÿÆEü‹G‰EÀ‰uÄÇEÈ    ÇEÌ  ÇEğ    ÆEüE¸ÿwMğj PèJŞÿ…Àˆî	  ‹uÜ‹†ø   ¾ø   …Àtö@û@u‹PüÑêëÿ0ÿp®[‹ĞÑêë3Ò3À9Eğu‹¶Aûƒà@¶EÂâ­+È‹jPRj WMğè <Şÿ‹ø…ÿˆ>	  j\Mğèìñ×ÿÿ5Ü$\Mğÿ5Ø$\è¨ñ×ÿEğPVè>"øÿ„À„ò  ÇEä    ÆEü‹Y‹Ê‹‚Ğ   …ÀEÈ‹h  ƒàƒÈ t ‹‹ÎÿP`= † tJ‹‹ÎÿÌ   ‹ğ…öuâë8‹Š,  ‹ÿP…Àt‹ğë%€    ‹‹ÎÿP`= † t‹‹ÎÿÌ   ‹ğ…öuâEä‹ÎPè§ú-‹ø…ÿyÆEü‹Eä…Àt‹PÿQéc  ÇEà    ÆEü
‹MÜ‹ÿPTEàPèúÿ‹ø…ÿy'ÆEü‹Eà…Àt‹PÿQÆEü‹Eä…Àt‹PÿQé  ‹Eä‹uàPH$QEğ‹ÎPè¬73‹ø…ÿy$ÆEü…öt‹VÿPÆEü‹Eä…Àt‹PÿQéÚ  ‹‹ËÿÜ   ‹H8‹ÿä   …À„  ‹‹ËÿÜ   ‹H8‹ÿä   MğQP‹Îè”f4‹ø…ÿy$ÆEü…öt‹VÿPÆEü‹Eä…Àt‹PÿQér  ‹‹ËÿÜ   ‹H8‹ÿä   j j jP‹Îè'Ä2‹ø…ÿy$ÆEü…öt‹VÿPÆEü‹Eä…Àt‹PÿQé%  ‹‹ËÿÜ   ‹H8‹ÿä   j jjP‹ÎèÚÃ2‹ø…ÿy$ÆEü…öt‹VÿPÆEü‹Eä…Àt‹PÿQéØ  ‹‹ËÿÜ   ‹H8‹ÿä   P‹Îè3w2‹ø…ÿy$ÆEü…öt‹VÿPÆEü‹Eä…Àt‹PÿQé‘  ‹‹ËÿÜ   ‹H8‹ÿä   ‹‹‘ø   MğQ‹ÈÿÒ‹ø…ÿy$ÆEü…öt‹VÿPÆEü‹Eä…Àt‹PÿQéB  3ÿÆEü…öt‹VÿPÆEü‹Eä…Àt‹PÿQé  èï0ÿÿ„Àu3ÿé  j Mğèš=Şÿ…Àˆ\  ‹K,EğPèFØıÿ…ÀˆN  ‹Eğ…À„i  ö@û@u‹Hüë
ÿ0ÿp®[‹È‹Eğ÷Áşÿÿÿ†E  …Àtö@û`u‹øë‹8ë3ÿ…Àtö@û@u‹@üÑèëÿ0ÿp®[Ñèë3À·DGşƒø/t'ƒø\t"ƒìMğ‹ÄÇ ğÑTÇ@   èƒ.Şÿ…ÀˆÁ  ‹
à$\ƒì‹ä$\‹Ä‰Mğ‰Pè].Şÿ…Àˆ¡  EğPVèKøÿ„À„  ‹‹ËÿĞ   ‹Ë‹°Ì   ‹ÿPH‹…Ét
öAû`t‹	ë3É‹Eğ…Àt
ö@û`t‹ ë3ÀhÊ  hø.]hØ9]VQPhø9]jj è×—İÿƒÄ$ÇEä    ÆEü'‹Ë‹ÿÔ   j MèQ‹H8è!j÷ÿÆEü(‹MèƒÁ‹‹P‹C(ÿpÿÒ‹ø…ÿy'ÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQéU  ‹MèUğƒÁR‹ÿP‹ø…ÿy'ÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQé  ‹EèÇ@(   Ç@,    ‹EèÇ@0   Ç@4    ‹EèÇ@8   Ç@<    ‹MèƒÁ‹ÿP<‹ø…ÿy'ÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQé®  ÇEÜ    ÆEü<èv‚ßÿ‹K(Pj j j ÿqEÜPèPúÿ‹ø…ÿy8ÆEü(‹EÜ…Àt‹PÿQÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQéJ  ‹uÜ‹Îj j è„çI‹ø…ÿy5ÆEü(…öt‹VÿPÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQé  ÇEà    ÆEüFEà‹Mèj j j PA÷ÙhÄ É#ÈQ‹Îè«÷I‹ø…ÿyFÆEü<‹Eà…Àt‹PÿQÆEü(…öt‹VÿPÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQé‡  ‹MàEäPèƒI‹ø…ÿyFÆEü<‹Eà…Àt‹PÿQÆEü(…öt‹VÿPÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQé/  ‹MèƒÁ‹ÿP@‹ø…ÿyFÆEü<‹Eà…Àt‹PÿQÆEü(…öt‹VÿPÆEü'‹Eè…Àt‹PÿQÆEü‹Eä…Àt‹PÿQéØ  ‹Eè…Àt‹PÿQÇEè    ÆEü<‹Eà…Àt‹PÿQÆEü(…öt‹VÿPÆEü'‹Eè…Àt‹PÿQ‹‹ËÿĞ   ‹ˆÌ   ‹Eä;H0…¦   ‹@\3ÿ‰EÜ…ÀS  wì‹‹ËÿPH‹…Ét
öAû`t‹	ë3É‹EäƒÿsƒÀxë‹@XÆ‹ ‹@0…Àt
ö@û`t‹ ë3Àhğ  hø.]hØ9]QPh :]jj è”İÿ‹EäƒÄ ƒÿsƒÀxë‹@XÆ‹Ë‹ ƒÀ0PèÌ:ûÿGƒÆ;}ÜŒrÿÿÿé½   ‹‹ËÿĞ   ‹ˆÌ   ‹Eä;H4…0  ‹@<3ÿ‰EÜ…À‘   wì‹‹ËÿPH‹…Ét
öAû`t‹	ë3É‹EäƒÿsƒÀXë‹@8Æ‹ ‹@0…Àt
ö@û`t‹ ë3Àhú  hø.]hØ9]QPh :]jj è\“İÿ‹EäƒÄ ƒÿsƒÀXë‹@8Æ‹Ë‹ ƒÀ0Pè
:ûÿGƒÆ;}ÜŒrÿÿÿÆEü‹Eä…Àt‹PÿQ3ÿë ÆEüMğj èj2ŞÿÆEü M¬èN}ŞÿÇEüÿÿÿÿ‹MìÿuØ‹I‹ÿPD‹Ç‹Môd‰
    Y_^[‹å]Ãè`œİÿPèÊ…İÿh¤9]h¯  hØ,]è††İÿPè°…İÿPèª…İÿPè¤…İÿPè…İÿh¤9]hÿ  hØ,]èZ†İÿh@XVjbhØWVèI†İÿÌÌÌÌÌÌÌÌÌfU‹ìjÿhÕFd¡    Pƒì@SVW¡ØŒ3ÅPEôd£    ‰eğ‹ù‰}ä‹_,‰]À‹ÿÔ   ‹@8‰EØ‹ó…Ût"@ ‹‹ÎÿP`=¢† t]‹‹ÎÿÌ   ‹ğ…öuâ‹
Y‹Ğ   …ÀEÈ‹h  ƒàƒÈ te‹ó…Û„—   ‹‹ÎÿP`= † „…   ‹‹ÎÿÌ   ‹ğ…öuŞës‹ó ‹‹ÎÿP`=¢† t‹‹ÎÿÌ   ‹ğ…öuâ‹†P  -t  taƒèvt\ëH¡Y‹ˆ,  ‹ÿP‹ğ…öu&‹ó…Ût f‹‹ÎÿP`= † t‹‹ÎÿÌ   ‹ğ…öuâ¾°  Æ  uhæ  è¼Lèÿ…À„Ë  ‹
Yè©ùèÿ„ÀuB‹5Yh   è•Lèÿƒøu€¾È¬  t‹G,¸Œ  ô  uÆEÜ ‹Ïè½/ÿÿ„Àt
ÆEàëÆEÜÆEà ÆEï EïP‹Ëèœ…÷ÿ‹ğ…öxt€}ï „€   ‹MØ‹ÿ   ‰EÌ‹‹A‰EÈ‹ƒX  ‰EØ‹ƒ\  ‰Eè»P  ê  u‹
Y‹±œ  ë‹sx‹‹ÏÿPPÿuàÿuÜÿuØÿuèVP‹MÌÿUÈ‹ğ…ö‰†  ‹Æ‹Môd‰
    Y_^[‹å]Â èg{ßÿ…Àtpèë3ö‹‹Îÿô  €¸G   …I  ‹‹Îÿô  ƒ¸  u
ƒ¸   „)  ‹  ÇEè    ÇEü   ‹UèRÿP‹ğ…öyÇEüÿÿÿÿ‹Eè…À„nÿÿÿ‹PÿQécÿÿÿè ëÿ‰E¸¡”W…Àt
ÿpÿøQT‹ğë3ö‰u¼…ötÿ†`  ÆEüÆEü‹}è‹‹@‰EÄ‹ƒX  ‰EÈ‹ƒ\  ‰EÌ»P  ê  u
¡Y‹€œ  ë‹Cx‰EØ‹Mä‹ÿPPÿuàÿuÜÿuÈÿuÌÿuØP‹ÏÿUÄ‰EÔÇEü   ëè–ôŞÿ‰EÔ¸#7o ÃÇEü   ‹]À‹EÔ‹u¼…Àtÿu¸èQ+ëÿÆEü…ötÿ`  ÇEüÿÿÿÿ‹Eè…Àt‹PÿQ‹}ä»P  ê  …S  èçyßÿ…ÀtUpè…ötN‹‹Îÿô  …Àt@‹‹Îÿô  ƒ¸  u
ƒ¸   „  ‹‹Îÿô  ƒ¸  u
ƒ¸   „õ   ‹
|«^d¡,   ‹4ˆ€¾X   uè0¤Qƒ¾    uCècyßÿ…Àt!ƒÀètèUyßÿ…ÀtHèë3É‹ÿô  …Àtÿu‹Ïèä‰üÿ‹ğ…öˆ¨ıÿÿéŒ   èàëÿ‹Ø‰]¸¡”W…Àt
ÿpÿøQT‹ğë3ö‰u¼…ötÿ†`  ÇEü   ÆEüÿu‹Ïè‘‰üÿ‰EĞÇEü   ë!è óŞÿ‰EĞ¸™8o ÃÇEü   ‹]¸‹EĞ‹u¼‹}ä…ÀtSèÚ)ëÿÇEüÿÿÿÿ…ötÿ`  ‹M…Ét
™<  ƒ;t"ë‹ÙÃ<  ‹Q‹ÏÿŒ  ‹ğ…öˆæüÿÿ‹M…Étƒ;u¸   ë3Àj PQ‹Ïè(ÿÿ‹ğ…öˆ¼üÿÿ3À‹Môd‰
    Y_^[‹å]Â ÌÌÌÌÌÌÌÌƒìV‹ñ‹D$‹H,‹F,‹€P  ;P  t=ê  u0ÿt$‹Îè ˆüÿ…Àx€|$ tj jÿj ‹ÎèÈvØÿ…Àx3À^ƒÄÂ ‹ÿPH‹Èèÿ¡×ÿPD$PèÙÿh@±
